### PR TITLE
[KARAF-3827] - Fix Cave rest service doesn't retrieve the repositories

### PR DESCRIPTION
**Pull request by fpapon**
Fix the response of **REST** service, add **JSON** response object.

The structure of the object have change : 

- **Repository** object tag with JAXRS annotations, attribute **name**, **location**, **increment** and a list of **Resource**

- **Resource** object tag with JAXRS annotations, attribute **identity**, **type**, **version**, **artefact** link to the XML repository elements.

Add dependency to the **org.apache.karaf.util.XmlUtils** to parse the XML repository.

Add QueryParam JAXRS annotations on the query param method

Change Produces JAXRS annotations **application/xml** to **appplication/json**

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @jbonofre

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java [hunk 1]
```diff
@@ -16,21 +16,145 @@
...
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
 
 @XmlRootElement(name = "cave-repository")
+@XmlAccessorType(XmlAccessType.NONE)
+@XmlType(name = "Repository", propOrder = {"name", "increment", "location", "resources"})
 public class Repository {
 
-    private final CaveRepository repository;
```
Comment (jbonofre) [inline server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java]: How do you access the actual `CaveRepository` now (I don't find the location) ?

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java [hunk 2]
```diff
@@ -47,9 +48,9 @@ public Service(CaveRepositoryService service) {
      * @throws Exception in case of creation failure.
      */
     @POST
-    @Consumes("application/xml")
-    @Produces("application/xml")
-    public Repository create(String name, boolean scan) throws Exception {
+    @Path("/repositories")
+    @Produces("application/json")
```
Comment (jbonofre) [inline server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java]: We annotation with JAXB but we provide json output. Thoughts ?

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java (hunk 1 above)
Comment (fpapon) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java]: I don't change the implementation of the method in the Service class :
```
public Repository getRepository(@PathParam("name") String name) {
        CaveRepository repository = service.getRepository(name);
        if (repository != null) {
            return new Repository(repository);
        } else {
            return null;
        }
}
```
repository service is passed in the constructor and used directly the retreive repository data :
```
public Repository(CaveRepository repository) {
        this.name = repository.getName();
        this.increment = repository.getIncrement();
        this.location = repository.getLocation();

        try {
Document doc = XmlUtils.parse(repository.getRepositoryXml().toExternalForm());
...
```

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java (hunk 2 above)
Comment (fpapon) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java]: I use annotation with JAXB to define and force attribute name in the json output like "cave-repositories" :
```
@XmlRootElement(name = "cave-repository")
@XmlAccessorType(XmlAccessType.NONE)
@XmlType(name = "Repository", propOrder = {"name", "increment", "location", "resources"})
public class Repository {
...
```
It's not mandatory but I wouldn't change existing object name return but we can change it.

Comment (fpapon) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java]: If I remove JABX annotations, I've got this error :

No message body writer has been found for class [Lorg.apache.karaf.cave.server.rest.Repository;, ContentType: application/json

Comment (fpapon) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java]: We need to add a JSONProvider to the Servlet implementing CFXNonSpringServlet :
```
        String alias = getString("cave.rest.alias", "/cave/rest");
        Hashtable<String, String> config = new Hashtable<>();
        if (getConfiguration() != null) {
            for (Enumeration<String> e = getConfiguration().keys(); e.hasMoreElements();) {
                String key = e.nextElement();
                String val = getConfiguration().get(key).toString();
                config.put(key, val);
            }
        }
        this.alias = alias;
        this.servlet = new CaveRestServlet(service);
        this.httpService.registerServlet(this.alias, this.servlet, config, null);
```
I've already done this with blueprint but not in a java way, I can try to do it.

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java (hunk 1 above)
Comment (jbonofre) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Repository.java]: Got it now. Thanks !

File: server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java (hunk 2 above)
Comment (jbonofre) [reply server/rest/src/main/java/org/apache/karaf/cave/server/rest/Service.java]: OK, so it uses JABX -> JSON converter. That's fine as soon as we support both. Thanks.

**Pull request closed without merging**